### PR TITLE
Add aten::median/aten::nanmedian and its variants

### PR DESCRIPTION
**Pull request by yucai-intel**
WIP

### CODE REVIEW

File: yaml/xpu_functions.yaml [hunk 1]
```diff
@@ -171,6 +171,8 @@ supported:
   - tanh
   - tanh_
   - tanh.out
+  - median
+  - median.out
```
Comment (majing921201) [inline yaml/xpu_functions.yaml:175]: no such schema in native_functions.yaml

File: src/ATen/native/xpu/Sorting.cpp [hunk 2]
```diff
@@ -75,4 +75,115 @@ Tensor XPUNativeFunctions::argsort(
...
+    int64_t k = (size - 1) / 2;
+    return at::where(sorted[-1].isnan(), sorted[-1], sorted[k]);
+  } else {
+    // For torch.nanmedian return the middle element among the non-nan values
+    int64_t k = ((size - 1) - sorted.isnan().sum().item<int64_t>()) / 2;
+    return sorted[k].clone(); // Clone so we aren't keeping `sorted` alive
+  }
+}
+
+std::tuple<Tensor&, Tensor&> XPUNativeFunctions::median_out(
```
Comment (majing921201) [inline src/ATen/native/xpu/Sorting.cpp:161]: miss registration in yaml

File: src/ATen/native/xpu/Sorting.cpp [hunk 3]
```diff
@@ -75,4 +75,115 @@ Tensor XPUNativeFunctions::argsort(
...
+    Tensor& indices) {
+  return median_with_indices_impl(
+      values, indices, self, dim, keepdim, /*ignore_nan=*/false);
+}
+
+Tensor XPUNativeFunctions::median(const Tensor& self) {
+  return median_impl(self, /*ignore_nan=*/false);
+}
+
+std::tuple<Tensor&, Tensor&> XPUNativeFunctions::nanmedian_out(
```
Comment (majing921201) [inline src/ATen/native/xpu/Sorting.cpp:175]: miss registration in yaml

File: src/ATen/native/xpu/Sorting.cpp [hunk 4]
```diff
@@ -75,4 +75,115 @@ Tensor XPUNativeFunctions::argsort(
...
+    const Tensor& self,
+    int64_t dim,
+    bool keepdim,
+    Tensor& values,
+    Tensor& indices) {
+  return median_with_indices_impl(
+      values, indices, self, dim, keepdim, /*ignore_nan=*/true);
+}
+
+Tensor XPUNativeFunctions::nanmedian(const Tensor& self) {
```
Comment (majing921201) [inline src/ATen/native/xpu/Sorting.cpp:185]: miss registration in yaml

File: src/ATen/native/xpu/sycl/SortingRadixSelect.h [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+  static inline RadixType convert(at::Half v) {
+    RadixType x = *((uint16_t*)&v);
+    RadixType mask = -((x >> 15)) | 0x8000;
+    return (x ^ mask);
+  }
+
+  static inline at::Half deconvert(RadixType v) {
+    RadixType mask = ((v >> 15) - 1) | 0x8000;
+    auto v_de = v ^ mask;
+    return *((at::Half*)&v_de);
```
Comment (majing921201) [inline src/ATen/native/xpu/sycl/SortingRadixSelect.h:132]: got warning here "warning: ‘v_de’ is used uninitialized [-Wuninitialized]"
I am considering replace with union

**Comment by fengyuan14** [comment]
Compare cpu
```
2024-07-30T05:05:36.7133148Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_median_xpu_bfloat16 [32mPASSED[0m[32m [ 36%][0m
2024-07-30T05:05:36.7149425Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_median_xpu_bool [33mSKIPPED[0m[32m [ 36%][0m
2024-07-30T05:05:36.7163592Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_median_xpu_complex128 [33mSKIPPED[0m[32m [ 36%][0m
2024-07-30T05:05:36.7177835Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_median_xpu_complex64 [33mSKIPPED[0m[32m [ 36%][0m
2024-07-30T05:05:36.7276699Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_median_xpu_float16 [32mPASSED[0m[32m [ 36%][0m
2024-07-30T05:05:36.7356347Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_median_xpu_float32 [32mPASSED[0m[32m [ 36%][0m
2024-07-30T05:05:36.7450477Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_median_xpu_float64 [32mPASSED[0m[32m [ 36%][0m

2024-07-30T05:05:38.5800806Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_nanmedian_xpu_bfloat16 [32mPASSED[0m[32m [ 38%][0m
2024-07-30T05:05:38.5815663Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_nanmedian_xpu_bool [33mSKIPPED[0m[32m [ 38%][0m
2024-07-30T05:05:38.5830158Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_nanmedian_xpu_complex128 [33mSKIPPED[0m[32m [ 38%][0m
2024-07-30T05:05:38.5844458Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_nanmedian_xpu_complex64 [33mSKIPPED[0m[32m [ 38%][0m
2024-07-30T05:05:38.5925127Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_nanmedian_xpu_float16 [32mPASSED[0m[32m [ 38%][0m
2024-07-30T05:05:38.6003464Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_nanmedian_xpu_float32 [32mPASSED[0m[32m [ 38%][0m
2024-07-30T05:05:38.6083577Z test_ops_xpu.py::TestCommonXPU::test_compare_cpu_nanmedian_xpu_float64 [32mPASSED[0m[32m [ 38%][0m
```

Reduction semantics
```
2024-07-30T05:33:01.2067301Z test_reductions_xpu.py::TestReductionsXPU::test_median_corner_cases_xpu [32mPASSED[0m[32m [ 19%][0m
2024-07-30T05:33:01.2591684Z test_reductions_xpu.py::TestReductionsXPU::test_median_nan_values_xpu_float16 [32mPASSED[0m[32m [ 19%][0m
2024-07-30T05:33:01.3029753Z test_reductions_xpu.py::TestReductionsXPU::test_median_nan_values_xpu_float32 [32mPASSED[0m[32m [ 19%][0m
2024-07-30T05:33:01.3523377Z test_reductions_xpu.py::TestReductionsXPU::test_median_nan_values_xpu_float64 [32mPASSED[0m[32m [ 19%][0m
2024-07-30T05:33:01.3715395Z test_reductions_xpu.py::TestReductionsXPU::test_median_real_values_xpu_float16 [32mPASSED[0m[32m [ 19%][0m
2024-07-30T05:33:01.3903506Z test_reductions_xpu.py::TestReductionsXPU::test_median_real_values_xpu_float32 [32mPASSED[0m[32m [ 19%][0m
2024-07-30T05:33:01.4096612Z test_reductions_xpu.py::TestReductionsXPU::test_median_real_values_xpu_float64 [32mPASSED[0m[32m [ 19%][0m
2024-07-30T05:33:01.4230123Z test_reductions_xpu.py::TestReductionsXPU::test_median_real_values_xpu_int32 [32mPASSED[0m[32m [ 19%][0m
2024-07-30T05:33:01.4375617Z test_reductions_xpu.py::TestReductionsXPU::test_median_real_values_xpu_int64 [32mPASSED[0m[32m [ 19%][0m
```

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**